packages/vite/src/node/server/middlewares/static.ts

```diff
@@ -84,6 +90,18 @@ export function serveRawFsMiddleware(): Connect.NextHandleFunction {
     // searching based from fs root.
     if (url.startsWith(FS_PREFIX)) {
       url = url.slice(FS_PREFIX.length)
+
+      // restrict files outside of `fsServeRoot`
+      if (
+        path
+          .relative(root, path.isAbsolute(url) ? url : `/${url}`)
+          .startsWith('../')
+      ) {
+        res.statusCode = 403
+        res.end()
+        return
+      }
+
```

Comment:
I tested in Windows, and it is failing. There are a few utils that are used to avoid issues with difference from posix. Using `fsPathFromId` and `normalizePath` from `utils.ts` should make this work across systems. I tested this for Windows
```suggestion
      // restrict files outside of `fsServeRoot`
      if (
        normalizePath(path.relative(root, fsPathFromId(url))).startsWith('../')
      ) {
        res.statusCode = 403
        res.end()
        return
      }

      url = url.slice(FS_PREFIX.length)
```